pkg/services/clouds/maas_service.go

```diff
@@ -0,0 +1,117 @@
+package clouds
+
+import (
+	"fmt"
+	"os"
+
+	maasclient "github.com/canonical/gomaasclient/client"
+	"github.com/canonical/gomaasclient/entity"
+	"github.com/spectrocloud-labs/prompts-tui/prompts"
+	"github.com/validator-labs/validatorctl/pkg/components"
+)
+
+var (
+	// GetMaasClient is defined to enable monkeypatching during testing
+	GetMaasClient = getMaasClient
+	host          = "https://maas.io/MAAS"
+)
+
+// ReadMaasClientProps gathers and validates MAAS client credentials
+func ReadMaasClientProps(c *components.MaasPluginConfig) error {
+	var err error
+	c.MaasAPIToken, err = prompts.ReadPassword("MAAS API token", c.MaasAPIToken, false, -1)
+	if err != nil {
+		return fmt.Errorf("failed to prompt for password for MAAS API token: %w", err)
+	}
+
+	if c.Validator.Host != "" {
+		host = c.Validator.Host
+	}
+	c.Validator.Host, err = prompts.ReadText("MAAS Domain", host, false, -1)
+	if err != nil {
+		return err
+	}
+
+	if err := validateMaasClient(c.Validator.Host, c.MaasAPIToken); err != nil {
+		val, err := handleMaasClientError(err)
+		if err != nil {
+			return err
+		}
+		if val == "Continue" {
+			return ReadMaasClientProps(c)
+		}
+		os.Exit(0)
+	}
+
+	return nil
+}
+
+func validateMaasClient(maasURL, maasToken string) error {
+	client, err := GetMaasClient(maasURL, maasToken)
+	if err != nil {
+		return err
+	}
+	// gomaasclient doesnt provide a direct way to validate, so we manually check that requests return success
+	_, err = client.Account.ListAuthorisationTokens()
+	if err != nil {
+		return err
+	}
+	if client == nil {
+		return err
+	}
+	return nil
+}
+
+func handleMaasClientError(err error) (string, error) {
+	errMsg := fmt.Sprintf("MAAS credentials validation failed with error: %v. Please update your credentials.", err)
+	val, err := prompts.Select(errMsg, []string{"Continue", "Exit"})
+	if err != nil {
+		return "", err
+	}
+	return val, nil
+}
+
+// GetMaasResourceProps fetches a list of resource pools and availability zones in the cluster
+func GetMaasResourceProps(c *components.MaasPluginConfig) ([]string, []string, error) {
```

Comment:
nit: it would be cleaner to have separate functions for retrieving either AZs or resource pools vs. lumping them together. Also, it's more idiomatic to return `nil`, not `[]string{}` unless there's a good reason to return an empty string slice on error.